File: courses/lesson4_1recycle_view_contacts/src/main/java/by/academy/lesson4/part1/contacts/DataItem.java
package by.academy.lesson4.part1.contacts;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Objects;

public class DataItem implements Parcelable{

    private String title;
    private String contact;
    private final boolean contactMail;

    public DataItem(String title, String contact) {
        this(title, contact, contact.contains("@"));
    }

    public DataItem(String title, String contact, boolean contactMail) {
        this.title = title;
        this.contact = contact;
        this.contactMail = contactMail;
    }

    protected DataItem(Parcel in) {
        title = in.readString();
        contact = in.readString();
        contactMail = in.readByte() != 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(title);
        dest.writeString(contact);
        dest.writeByte((byte) (contactMail ? 1 : 0));
    }

    public static final Creator<DataItem> CREATOR = new Creator<DataItem>() {
        @Override
        public DataItem createFromParcel(Parcel in) {
            return new DataItem(in);
        }

        @Override
        public DataItem[] newArray(int size) {
            return new DataItem[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setContact(String contact) {
        this.contact = contact;
    }

    public String getContact() {
        return contact;
    }

    public boolean isContactMail() {
        return contactMail;
    }

    public String getTitle() {
        return title;
    }

    public void copyFrom(DataItem item) {
        this.contact = item.contact;
        this.title = item.title;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DataItem dataItem = (DataItem) o;
        return title.equals(dataItem.title) &&
                contact.equals(dataItem.contact);
    }

    @Override
    public int hashCode() {
        return Objects.hash(title, contact);
    }
}

File: courses/lesson5_cars/src/main/java/by/academy/lesson5/cars/DataItem.kt
package by.academy.lesson5.cars

import android.os.Parcel
import android.os.Parcelable
import java.util.*

data class DataItem(
        var ownerName: String?,
        var producer: String?,
        var model: String?,
        var plateNumber: String?,

) : Parcelable {
    constructor(parcel: Parcel) : this(
            parcel.readString(),
            parcel.readString(),
            parcel.readString(),
            parcel.readString()) {
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(ownerName)
        parcel.writeString(producer)
        parcel.writeString(model)
        parcel.writeString(plateNumber)
    }

    override fun describeContents(): Int {
        return 0
    }

    fun copyFrom(item: DataItem) {
        this.ownerName = item.ownerName;
        this.producer = item.producer;
        this.model = item.model;
        this.plateNumber = item.plateNumber;
    }


    companion object CREATOR : Parcelable.Creator<DataItem> {
        override fun createFromParcel(parcel: Parcel): DataItem {
            return DataItem(parcel)
        }

        override fun newArray(size: Int): Array<DataItem?> {
            return arrayOfNulls(size)
        }
    }


}